hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/MergeOperation.java

```diff
@@ -25,74 +25,107 @@
-import static com.hazelcast.util.MapUtil.createHashMap;
+import static com.hazelcast.spi.impl.merge.MergingValueFactory.createMergingValue;
 
 /**
- * Contains multiple merge entries for split-brain healing with a {@link SplitBrainMergePolicy}.
+ * Contains an entire ringbuffer for split-brain healing with a
+ * {@link SplitBrainMergePolicy}.
  *
  * @since 3.10
  */
-public class MergeOperation extends Operation implements IdentifiedDataSerializable, BackupAwareOperation, ServiceNamespaceAware {
+public class MergeOperation extends Operation
+        implements IdentifiedDataSerializable, BackupAwareOperation, ServiceNamespaceAware {
 
     private ObjectNamespace namespace;
-    private SplitBrainMergePolicy<Object, RingbufferMergeTypes> mergePolicy;
-    private List<RingbufferMergeTypes> mergingEntries;
+    private SplitBrainMergePolicy<RingbufferMergeData<Object>, RingbufferMergeTypes> mergePolicy;
+    private Ringbuffer<Object> mergingRingbuffer;
 
+    private transient Ringbuffer<Object> resultRingbuffer;
     private transient RingbufferConfig config;
-    private transient RingbufferContainer<Object, Object> ringbuffer;
-    private transient Map<Long, Data> valueMap;
 
     public MergeOperation() {
     }
 
-    public MergeOperation(ObjectNamespace namespace, SplitBrainMergePolicy<Object, RingbufferMergeTypes> mergePolicy,
-                          List<RingbufferMergeTypes> mergingEntries) {
+    public MergeOperation(ObjectNamespace namespace,
+                          SplitBrainMergePolicy<RingbufferMergeData<Object>, RingbufferMergeTypes> mergePolicy,
+                          Ringbuffer<Object> mergingRingbuffer) {
         this.namespace = namespace;
         this.mergePolicy = mergePolicy;
-        this.mergingEntries = mergingEntries;
+        this.mergingRingbuffer = mergingRingbuffer;
     }
 
     @Override
     public void beforeRun() throws Exception {
-        RingbufferService service = getService();
+        final RingbufferService service = getService();
         config = getRingbufferConfig(service, namespace);
-        ringbuffer = service.getOrCreateContainer(getPartitionId(), namespace, config);
     }
 
     @Override
     public void run() throws Exception {
-        valueMap = createHashMap(mergingEntries.size());
-        for (RingbufferMergeTypes mergingEntry : mergingEntries) {
-            long resultSequence = ringbuffer.merge(mergingEntry, mergePolicy);
-            if (resultSequence != -1) {
-                valueMap.put(resultSequence, ringbuffer.readAsData(resultSequence));
+        final RingbufferService service = getService();
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
+        RingbufferContainer<Object, Object> existingContainer = service.getContainerOrNull(getPartitionId(), namespace);
+
+        final RingbufferMergeTypes mergingValue = getMergingType(mergingRingbuffer);
```

Comment:
This is the only merge operation which contains the merge logic, instead of delegating it to the container. I guess it's due to the `service.destroyContainer(getPartitionId(), namespace);` and the `RingbufferContainer` not having a reference to its service.

Reply:
Yes, I'll probably change this to get in line with your current IQueue merge PR

Reply:
This comment is outdated by now. Most of the data structures have no reference to the service in their container. Eventually there will be more data structures where the merge logic will be in the operation. So it looks like this is the way to go.